build.py

```diff
@@ -0,0 +1,357 @@
+        if not action.option_strings:
+            [result] = self._metavar_formatter(action, action.dest)(1)
+        else:
+            parts = []
+            if action.nargs == 0:
+                parts.extend(action.option_strings)
+            else:
+                default = action.dest.upper()
+                args_string = self._format_args(action, default)
+                for option_string in action.option_strings:
+                    parts.append(str(option_string))
+                parts[-1] += ' ' + str(args_string)
+            result = ', '.join(parts)
+        return result
+
+def main(program, *args):
+    program_name = os.path.basename(program)
+    program_dir = os.path.dirname(os.path.realpath(__file__))
+    argparser = argparse.ArgumentParser(prog=program_name, formatter_class=SingleMetaVarHelpFormatter)
+    argparser.add_argument('-l', '--language', default='python', metavar='LANGUAGES', help=
+        "A comma-separated list of languages to build native libraries %(dest)s.\n"
+        "Supported languages include \"python\" and \"java\" (default: %(default)s).")
+    argparser.add_argument('-p', '--python', default=None, help=
+        "which python executable\n(default: PYTHON_BIN_PATH or current interpreter)")
+    parsed_args = argparser.parse_args(args)
+    env = os.environ.copy()
+    use_current_python_interpreter = False
+    if parsed_args.python is not None:
+        env['PYTHON_BIN_PATH'] = parsed_args.python
+    elif 'PYTHON_BIN_PATH' not in env:
+        env.setdefault('PYTHON_BIN_PATH', os.path.normpath(get_current_python_cli_executable()))
+        use_current_python_interpreter = True
+    pyarrow_install = not os.getenv('SKIP_PYARROW_INSTALL') and sys.platform != 'win32'
+    command_args = ['bazel', 'build']
+    for language in parsed_args.language.split(','):
+        if not language: continue
+        if language.startswith('//') or language.startswith('@'):
+            command_args.append(language)
+        elif 'python' in language:
+            ray_build_python = command_args.append('//:ray_pkg')
+        elif 'java' in language:
+            ray_build_java = command_args.append('//java:all')
+        else: raise ValueError("Unrecognized language: " + language)
+    result = 0
+    try:
+        # The following line installs pyarrow from S3, these wheels have been
+        # generated from https://github.com/ray-project/arrow-build from
+        # the commit listed in the command.
+        if pyarrow_install and result == 0:
+            pip_args = ('install', '-q',
+                '--target=' + os.path.join(program_dir, "python", "ray", "pyarrow_files"), 'pyarrow==0.14.0.RAY',
+                '--find-links', "https://s3-us-west-2.amazonaws.com/arrow-wheels/516e15028091b5e287200b5df77d77f72d9a6c9a/index.html")
+            result = pip_main(list(pip_args)) if use_current_python_interpreter else proc_call((env['PYTHON_BIN_PATH'], '-m', 'pip') + tuple(pip_args), check=False)[0]
```

Comment:
It is better to call subprocess.Popen (or checked_call) here I think, similar to how it is done in the shell script at the moment.

Reply:
Any thoughts on the comment I put in `pip_main`? I explained the reason for this choice there.

Reply:
Any thoughts on the comment I put in `pip_main`? I explained the reason for this choice there.